### PR TITLE
ForwardRef support for TextField

### PR DESCRIPTION
**Pull request by reesscot**
*Description of changes:*
This PR adds forward ref support for the `TextField` primitive by wrapping it in `React.forwardRef`. It also required some TS magic to infer the correct HTML type based on the `isMultiline` prop. 

```jsx
const CustomerComponent = () => {
  const ref = React.useRef(null);
  
  // Below returns INPUT because the ref is forwarded down to DOM element (TextField => Input => <input>)
  // ref.current.nodeName
  
  return (
    <TextField label="name" ref={ref} />
  );
};
```

After this PR is merged I will be able to add forwardRef support to `PasswordField` and other `TextField` based primitives.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by hvergara — APPROVED** [review]
LGTM

File: packages/react/src/primitives/TextField/TextField.tsx [hunk 1]
```diff
@@ -110,4 +108,16 @@ export const TextField = <Multiline extends boolean>(
   );
 };
 
-TextField.displayName = 'TextField';
+export const TextField = React.forwardRef(TextFieldPrimitive) as <
```
Comment (zchenwei) [inline packages/react/src/primitives/TextField/TextField.tsx:111]: This seems dissatisfying but don't have a good a idea on my mind

**Review by zchenwei — APPROVED** [review]
LGTM 🚢

File: packages/react/src/primitives/TextField/TextField.tsx (hunk 1 above)
Comment (reesscot) [reply packages/react/src/primitives/TextField/TextField.tsx:111]: Yeah, I tried a number of different solutions, but the other options created new problems. This was the only option that fully worked. Perhaps we can revisit it later.

Comment (reesscot) [reply packages/react/src/primitives/TextField/TextField.tsx:111]: For posterity, here is one of the other solutions I tried, which involves redeclaring the forwardRef function definition.

```jsx
// Redeclare `React.forwardRef` to support higher order function type inference
// which only works with plain function types (default definition )
// see: https://fettblog.eu/typescript-react-generic-forward-refs/
//
// Note: Without this only shared props between InputProps and TextAreaProps will be available, a
// and any unique props such as `resize` on TextArea or HTMl attributes will be missing from the type.
declare module 'react' {
  function forwardRef<T, P = {}>(
    render: (props: P, ref: React.Ref<T>) => React.ReactElement | null
  ): (props: P & React.RefAttributes<T>) => React.ReactElement | null;
}
```

The problem with this approach is that it affected other modules causing them to not be able to have a `displayName`. 

Ref: https://fettblog.eu/typescript-react-generic-forward-refs/

**Pull request merged**